x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java

```diff
@@ -662,163 +661,261 @@ public void storeModel(Model model, ActionListener<Boolean> listener, TimeValue
+                // if we got here that means the configuration store bulk item succeeded, so we did modify an index
+                responses.add(new StoreResponseWithIndexInfo(secretsStoreResponse, true));
+            } else {
+                responses.add(new StoreResponseWithIndexInfo(configStoreResponse, true));
+                if (modelFromBulkItem != null) {
+                    successfullyStoredModels.add(modelFromBulkItem);
+                }
+            }
+        }
+
+        return new ResponseInfo(responses, successfullyStoredModels);
     }
 
-    private static void logBulkFailure(String inferenceEntityId, BulkItemResponse item) {
-        if (item.isFailed()) {
+    private static ModelStoreResponse createModelStoreResponse(BulkItemResponse item, Map<String, String> docIdToInferenceId) {
+        var failure = item.getFailure();
+
+        String inferenceIdOrUnknown = "unknown";
+        var inferenceIdMaybeNull = docIdToInferenceId.get(item.getId());
+        if (inferenceIdMaybeNull == null) {
+            logger.warn("Failed to find inference id for document id [{}]", item.getId());
+        } else {
+            inferenceIdOrUnknown = inferenceIdMaybeNull;
+        }
+
+        if (item.isFailed() && failure != null) {
             logger.warn(
-                format("Failed to store inference endpoint [%s] index: [%s]", inferenceEntityId, item.getIndex()),
-                item.getFailure().getCause()
+                format(
+                    "Failed to store document id: [%s] inference id: [%s] index: [%s] bulk failure message [%s]",
+                    item.getId(),
+                    inferenceIdOrUnknown,
+                    item.getIndex(),
+                    item.getFailureMessage()
+                )
             );
+
+            return new ModelStoreResponse(inferenceIdOrUnknown, item.status(), failure.getCause());
+        } else {
+            return new ModelStoreResponse(inferenceIdOrUnknown, item.status(), null);
         }
     }
 
-    private static BulkItemResponse.Failure getFirstBulkFailure(BulkResponse bulkResponse) {
-        for (BulkItemResponse item : bulkResponse.getItems()) {
-            if (item.isFailed()) {
-                return item.getFailure();
-            }
+    private static Model getModelFromMap(@Nullable String inferenceId, Map<String, Model> inferenceIdToModel) {
+        if (inferenceId != null) {
+            return inferenceIdToModel.get(inferenceId);
         }
 
         return null;
     }
 
+    private void updateClusterState(List<Model> models, ActionListener<AcknowledgedResponse> listener, TimeValue timeout) {
+        var inferenceIdsSet = models.stream().map(Model::getInferenceEntityId).collect(Collectors.toSet());
+        var storeListener = listener.delegateResponse((delegate, exc) -> {
+            logger.warn(format("Failed to add inference endpoint %s minimal service settings to cluster state", inferenceIdsSet), exc);
```

Comment:
This might be better as "Failed to add minimal service settings to cluster state for inference endpoints %s"